fgpyo/sam/__init__.py

```diff
@@ -1036,55 +1043,119 @@ class ReadEditInfo:
+            including ambiguity codes like R and N).
         reference_offset: if provided, assume that reference_sequence[reference_offset] is the
             first base aligned to in reference_sequence, otherwise use r.reference_start
 
     Returns:
         a ReadEditInfo with information about how the read differs from the reference
     """
-    assert not rec.is_unmapped, f"Cannot calculate edit info for unmapped read: {rec}"
+    if rec.is_unmapped or rec.query_sequence is None or rec.query_sequence == NO_QUERY_BASES:
+        return None
 
-    query_offset = 0
-    target_offset = reference_offset if reference_offset is not None else rec.reference_start
-    cigar = Cigar.from_cigartuples(rec.cigartuples)
+    query_offset: int = 0
+    target_offset: int = reference_offset if reference_offset is not None else rec.reference_start
+    cigar: Cigar = Cigar.from_cigartuples(rec.cigartuples)
 
     matches, mms, insertions, ins_bases, deletions, del_bases = 0, 0, 0, 0, 0, 0
-    ok_bases = {"A", "C", "G", "T"}
-
+    md_edits: list[str] = []
+    current_match_count: int = 0
     for elem in cigar.elements:
         op = elem.operator
-
-        if op == CigarOp.I:
-            insertions += 1
-            ins_bases += elem.length
-        elif op == CigarOp.D:
+        # TODO: use match-case statements after we drop Python 3.9 support
+        if op in (CigarOp.M, CigarOp.X, CigarOp.EQ):
+            for in_block_offset in range(0, elem.length):
+                if (target_offset + in_block_offset) >= len(reference_sequence):
+                    break  # out of bounds
+                query_base: str = rec.query_sequence[query_offset + in_block_offset].upper()
+                ref_base: str = reference_sequence[target_offset + in_block_offset].upper()
+                is_match = _is_match_base(
+                    query_base=query_base, ref_base=ref_base, match_htsjdk=match_htsjdk
+                )
+
+                if is_match:
+                    matches += 1
+                    current_match_count += 1
+                else:  # mismatch
+                    md_edits.append(str(current_match_count))  # append match count and reset
+                    current_match_count = 0
+                    md_edits.append(ref_base)  # grab mismatched base from the reference
+                    mms += 1
+            query_offset += elem.length_on_query
+            target_offset += elem.length_on_target
+        elif op == CigarOp.D:  # consumes ref
+            md_edits.append(str(current_match_count))  # append match count and reset
+            md_edits.append("^")
+            for in_block_offset in range(0, elem.length):
+                if (target_offset + in_block_offset) >= len(reference_sequence):
+                    break  # out of bounds
+            md_edits.append(f"{reference_sequence[target_offset:target_offset + elem.length]}")
```

Comment:
chore: (1) I think we need to upper case the reference and (2) I don't think we need an f-string since reference sequence is already a string
```suggestion
            md_edits.append(reference_sequence[target_offset:target_offset + elem.length].upper())
```